src/core/zipfile.cpp

```diff
@@ -52,11 +76,20 @@ void ZipFile::Read(void *data, size_t size) {
 	z.avail_out = size;
 	do {
 		if (!z.avail_in) {
-			z.next_in = reinterpret_cast<Bytef*>(&buffer[0]);
-			z.avail_in = file.Read(&buffer[0], buffer.size());
+			if (is_file) {
+				z.next_in = reinterpret_cast<Bytef*>(&buffer[0]);
+				z.avail_in = file.Read(&buffer[0], buffer.size());
+			} else {
+				z.next_in = reinterpret_cast<Bytef*>(&index_buffer[0]);
```

Comment:
Having this inside the read loop doesn't really make any sense, since it's not actually consuming the input and so implicitly relies on the loop only running once. I think pulling the non-file init of next_in/avail_in out of the loop would make what's happening clearer.